Add Multiple Changes per Block Position Capabilities

gabizou opened the pull request:
This is an overarching fix for adding the ability to properly track block changes from things like pistons by exposing internal intermediary transactions of block changes to positions. TLDR is at the bottom.

##### Ok, but why?

To understand the why, you need to get a little background. Currently, the [`PhaseTracker`](https://github.com/SpongePowered/SpongeCommon/blob/198e8ce33c40adb86e79d707e8d58f29a607f1e3/src/main/java/org/spongepowered/common/event/tracking/TrackingUtil.java#L336) will capture blocks via a [`CaptureListSupplier`](https://github.com/SpongePowered/SpongeCommon/blob/198e8ce33c40adb86e79d707e8d58f29a607f1e3/src/main/java/org/spongepowered/common/event/tracking/context/CapturedBlocksSupplier.java) object, which is pretty much a wrapper around a lazily initialized `List` of `BlockSnapshot`s. This is all fine and dandy when you consider most things happening in the game will be one-off or multiple placement transactions, but for things like Pistons, they perform a lot of operations involving block changes, tile entity replacements, and individualized neighbor notifications after blocks and tiles have been moved/placed. What ends up happening at this point is that the `List` will have duplicated block positions (read: multiple `BlockSnapshot`s and therefor, multiple [`Transaction<BlockSnapshot>`s](https://github.com/SpongePowered/SpongeCommon/blob/198e8ce33c40adb86e79d707e8d58f29a607f1e3/src/main/java/org/spongepowered/common/event/tracking/TrackingUtil.java#L468) to exist, and worst yet, they will only understand the *original* block change, not the "final" change.

##### I guess, but why is that a problem?

It's a problem because take this situation: A Piston is going to move. What does it do?
- Iterates over all blocks it needs to destroy
  - Tells the block to drop itself as items (snow drops snowballs)
  - Sets the block to air with no notifications
- Iterates over all blocks it needs to move
  - Sets the original position to air with no notifications
  - Offsets to the facing direction
  - Sets the offsetPosition to a Piston Extension block with no notifications
  - Sets a TileEntity for moving piston on the offset position
- If the piston is extending
  - Sets the "piston head" piston extension block with no notifications
  - Sets the piston tile entity
- Iterates in *reverse order* destroyed blocks first, then moved blocks
  - Sends a neighbor notification of the original blocks that were destroyed/moved
- If extending, sends a neighbor notification to the extended piston head

Now, this may be a little confusing, but, with the multiple blocks being moved, you can have a position potentially having **two** transactions! Let's explain:
1) Block being moved gets set to air
2) Block being moved gets set to Piston Extension
Respectively, this would be represented incorrectly as two different `Transaction`s and worst yet, because of the core logic with how snapshots are generated for these transactions, [the snapshot only gets a change flag assigned when the block change is captured](https://github.com/SpongePowered/SpongeCommon/blob/198e8ce33c40adb86e79d707e8d58f29a607f1e3/src/main/java/org/spongepowered/common/event/tracking/TrackingUtil.java#L346), so the first transaction will have a `BREAK`, and the second would get a `PLACE`, whereas we both know in reality, it's a `MODIFY` in cases where there was an existing block in place!

##### Huh, I guess... So, how does this PR fix that?

Welcome to my world... The few hundreds of lines being changed involve specifying certain `IPhaseState`s to selectively capture not just block changes, but also `TileEntity` changes, and neighbor notification requests. This got a little [messy in the later stages](https://github.com/SpongePowered/SpongeCommon/blob/e6e0e53042c88838d527ab464b64a7725ddce747/src/main/java/org/spongepowered/common/event/tracking/context/MultiBlockCaptureSupplier.java#L571-L622) especially for transaction application; however, it is properly capturing and playing back the physics that would otherwise have not been captured had we disabled capturing entirely.... For the most part.... What this does now is builds a singular `Transaction<BlockSnapshot>` only providing the *updated* change according to [the capture supplier having multiple changes for the desired position](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/event/tracking/context/MultiBlockCaptureSupplier.java#L90-L168) while still being lazily evaluated. This neat change will apply globally to all `IPhaseState`s that end up potentially capturing multiple changes to a single block position, but the advantage is that for the specific `IPhaseState`s that need more refined tracking (read: tile entity changes and neighbor notifications), that is added as a possible [tracking hook to `IPhaseState`](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/event/tracking/phase/tick/BlockEventTickPhaseState.java#L100-L118).

##### Why did you say "For the most part...."?

Well... Here's the catch.... Tracking is hard. The way that Sponge handles tracked block changes is by allowing block changes to pass through to the [`Chunk without making explicit calls to performing block physics`](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/mixin/core/world/MixinChunk.java#L602-L737). Because of that, there's potential cases where [played back notifications](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/event/tracking/context/BlockTransaction.java#L265-L272) or [`onBlockAdded` calls](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/event/tracking/context/BlockTransaction.java#L198-L228) can have different BlockStates than what currently exists on the `Chunk`, leading to vanilla logic to validate itself to setting a block to Air instead of performing the desired operation that it originally intended to perform. As much as I'd like to explain the depth of that logic, it comes down to this:
Because tracking injects itself to capture changes as transactions, but allows the Chunk to still get the block change, without playing any physics, and then plays back the "physics" after all of those changes to the target `Chunk`, we are slightly modifying the order of operations that ends up causing (as an example) [T-flop pistons](https://minecraft.gamepedia.com/Tutorials/Piston_circuits#T-Flip_Flops) to potentially break as the tile entity being processed at the time will see an invalid state and set itself to air... I'm working on [re-applying a systematic way to provide in-progress `IBlockState`s to the World to alleviate this issue](https://github.com/SpongePowered/SpongeCommon/blob/9906e5ef315ba70301bd8257677f84f410b96bd1/src/main/java/org/spongepowered/common/event/tracking/context/SpongeProxyBlockAccess.java#L37), but that is an experiment I'm currently working on.

#### So, it's complicated, I get it. I don't fully understand what or how this all works, but I guess you do... My question is, is it going to cause lag?

Maybe? From what I've seen, it's not causing any increase in lag from my limited testing. Once I've ~~called it quits to resolving the 1 tick piston issue~~ decided this branch is satisfactory with how it functions, I'll be reviewing performance comparisons with `stable-7` and sending out some test builds on discord. For the most part, this entire system was designed with minimal memory footprint additions as possible. Multi-block changes are not adding any new objects until it is detected a duplicate transaction is being added for a position, and even then, when that happens, the actual `BlockTransaction` bits aren't used until certain `IPhaseState`s decide it needs to use them. AND EVEN THEN, it's optimized in that it only stores references to objects that otherwise we'd need to be able to play back (process) or cancel (restore). Restores are working beautifully btw, but playback is the only issue at current writing.

#### TLDR; Previous tracking system doesn't know how to handle two block changes to a single position. This fixes it, and allows using ChangeBlockEvent to cancel pistons. It's complicated.

gabizou committed:
Lord of the Blocks: The Fellowship of the Tracker
Introduce BlockPosMultiMap. The context of its usage is to
replace the previous existing CaptureObject that maintained
a List of BlockSnapshots such that this new Multimap will
retain a set of BlockPoses to maintain a List of
intermediary BlockSnapshot changes for each BlockPos.

The reasoning for this is that certain case of phases will
have multiple changes per block position and require
processing of said positions. The issue is that while these
multiple changes need to be tracked, the capturing would
be excessive for cases where multiple changes per position
do NOT take place. This specific ICaptureSupplier is designed
to retain the same footprint of the original block capture
object, but with the advantage of dynamically changing over
to a Multimap based implementation when necessary.

The costs for processing and viewing are yet to be determined,
however the implementation is still needing to be updated for
the changes as processing needs to reflect the order in which
block changes take place, while still processing the
ChangeBlockEvent based on the original BlockSnapshot that existed
compared to the BlockSnapshot that is taking it's place.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Aaron1011 commented on src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java:
```diff
@@ -189,6 +189,8 @@ public void construct(World worldIn, GameProfile gameProfileIn, CallbackInfo ci)
         this.targetedLocation = VecHelper.toVector3d(worldIn.getSpawnPoint());
     }
 
+
```
These lines can be removed

Aaron1011 commented on src/main/java/org/spongepowered/common/event/tracking/context/BlockPosMultiMap.java:
```diff
@@ -0,0 +1,268 @@
+        return Collections.unmodifiableList(this.snapshots == null ? Collections.emptyList() : this.snapshots);
+    }
+
+    public final void prune(BlockSnapshot snapshot) {
```
Should there be a method body, or can this be removed?

liach replied:
guess this is still wip.

Aaron1011 commented on src/main/java/org/spongepowered/common/event/tracking/context/BlockPosMultiMap.java:
```diff
@@ -0,0 +1,268 @@
+public final class BlockPosMultiMap implements ICaptureSupplier {
+
+    @Nullable private ListMultimap<BlockPos, SpongeBlockSnapshot> multimap;
+    @Nullable private List<SpongeBlockSnapshot> snapshots;
```
Based on how this field is used, it looks like you could just initialize it here - none of the other methods ascribe any special meaning to it being `null`.

gabizou replied:
There's plenty of meaning for it to be `null` by default:
- No blocks captured? No wasted memory
- `isEmpty()` checks are blazing fast
- `put(BlockSnapshot)` is more fasts on initialization due to fail fast if statements on nullability
- Because there's thousands of these objects being created per tick, they need to construct and be as lightweight as possible.

Remember, a lot of the capture suppliers have had nullable fields for these reasons to lower the average memory cost of the PhaseTracker in general.

gabizou committed:
Lord of the Blocks: The Two Captures

Redirected all usages of TrackingUtil#processBlockCaptures to use
the new BlockPosMultiMap. Processing and pruning still needs to be
worked on, but the project will compile now.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

liach commented on src/main/java/org/spongepowered/common/event/tracking/context/BlockPosMultiMap.java:
```diff
@@ -0,0 +1,317 @@
+     * @return An unmodifiable list of first block originals being changed
+     */
+    public final List<SpongeBlockSnapshot> get() {
+        return Collections.unmodifiableList(this.snapshots == null ? Collections.emptyList() : this.snapshots);
```
The empty list is already unmodifiable, so
```java
return this.snapshots == null ? Collections.emptyList() : Collections.unmodifiableList(this.snapshots);
```

liach commented on src/main/java/org/spongepowered/common/event/tracking/phase/tick/BlockTickPhaseState.java:
```diff
@@ -116,9 +117,9 @@ public boolean doesCaptureEntitySpawns() {
-        if (context.shouldProcessImmediately()) {
-            TrackingUtil.processBlockCaptures(context.getCapturedBlocks(), this, context);
-            context.getCapturedBlockSupplier().get().remove(snapshot);
+        if (false) {
```
add todo?

gabizou commented:
Just to clarify, this *is* still a work in progress for implementation.

gabizou committed:
Lord of the Blocks: The Five Classes

Documenting the uses of PhaseContext#getCapturedBlocks vs using the
MultiBlockCaptureSupplier. Started refactoring the processing of captured
blocks such that TrackingUtil no longer uses a passed object that would
otherwise be retrieved from the PhaseContext being passed in to begin
with.

Added many todo's for handling individual block snapshot removal, and the
consideration of changes to be made for Transaction as a class.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Road to Intermediate Transactions

I believe this will work for processing of intermediary transactions
when multiple block changes are considered. Will be testing and modifying
as necessary. To test, a special test plugin will be written for
the next commit to verify piston cancellations will perform correctly.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Road to Eventgard

Actually implemented most of MultiBlockCapture being processed, but
finding that it may be necessary to capture notification events as well as
tile entity creation and removal during block events to properly play them
back for proper cancellation, avoiding client issues. This sort of cost
should be minimal, once optimized and verified it works.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Voice of Pistoman

Get part of block event tile entity tracking functional.

What this change does introduce is the ability to track
TileEntity modification on the Chunk, block physics performed
when TileEntities are affected during ticking themselves,
and the ability to play back notifications that were delayed
during BlockEvent processing. Vanilla functionality still
works in that neighbor notifications are being properly handled
and the remaining edits are still processed in the order in which they
were applied.

Currently, cancelling any ChangeBlockEvents with Pistons involved
will still cancel, but render a ghosted TileEntity remaining in the
world causing a disappearance of the blocks. Verification is required
for whether this is due to the Client performing a similar edit, or
whether the Server is in fact retaining a ghosted TileEntity.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Pistons of the Marshes

Migrated what was existing on the EnhancedPhaseContext to
MultiBlockCaptureSupplier. All of the code is optional only
to enable transaction tracking through the specific PhaseStates
that will utilize it. Unique and specific transactions that are
defined showing their tracked original snapshot allows us to
rollback correctly.

Rollbacks are still somewhat finicky, but it is very close to
being handled correctly.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Blockeventir

Merges upstream changes in stable-7 for neighbor notification
modifications, some block event modifications. Not all changes
are actually supported yet, need to rewire some of the additions
to avoid frames and context manipulation via hooks.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Taming of Eventgol

Most of this is cleanup post merging and clarifying some things
about using packet frames instead of constant frame manipulation in
phase states. There is still the question of verifying BlockEventQueue
phase being necessary. As it stands, it serves that the block event
being queued does lack enough information, but that information should
be passed by the current PhaseContext, and not needed to be stored
on a whole new PhaseContext and PhaseState.

More to come. Fixed a few bugs with invalid transactions still being
applied.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Logic Gate is Closed

Part of this is functional to the respect of being
able to cancel pistons fully, without issue of the client
being aware of the simulation. In near perfect replication
of cancelling the block changes from pistons performing
movements, without the added issue of requiring to listen to
ChangeBlockEvent.Pre with piston context keys. Only issue
is some odd notification propogation not taking place
after the fact.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: Of Blocks and Notified Neighbors

Rewired post processing to take advantage of the multimap of
EventTransactions retained by MultiBlockCaptureSupplier. This ought
to propogate notifications better, and more precisely, however, there
is still an issue of chained propogation not taking place that causes
odd behavior in complex piston machines with chained/timed practices
based on notifications from moved blocks after the fact. Will investigate
further.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Lord of the Blocks: The Neighbor on the Block

Refactored the special handling of per tile and neighbor transactions as
proper transactions. The side effect of these transactions are that they
need to be post applied without bulk capturing, or risking the duplication
and deeply nested transaction capturing that could prove to cause stack
overflows all over again. The flip of the "immediate" versus "normal" post
process is that the "normal" capturing will take place for all states that
are not considered requiring "special" processing (like BlockEvent, or
TileEntityTick).

So far, it's been verifiable that the entire transaction process is
proceeding semi-normally, with the exception of "instant flip" pistons
being awkwardly broken where the piston attempting to retract a block that
it potentially pushed will result in an "air" block being placed, and the
original block that was intended to be retracted with the sticky piston
will quite literally disappear. The reasoning for this I am assuming is
due to the nature of various getBlockState checks for Pistons as blocks
and TileEntities that will reset the blocks to "air" if they're not done
in the proper order.

The good thing is that cancellations still work without having the
required ChangeBlockEvent.Pre event listener with context key checks.
Another positive change is that the previous injections for Pistons
specifically to perform neighbor notifications are now inlined with the
appropriate IPhaseStates to be thrown without special casing.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou commented:
For those curious, I've updated the PR description explaining the original issue, and how this solves it, along with some side effects (and improvements).

Zidane commented on src/main/java/org/spongepowered/common/event/tracking/PhaseTracker.java:
```diff
@@ -604,31 +609,45 @@ public void notifyBlockOfStateChange(final IMixinWorldServer mixinWorld, final B
+    @SuppressWarnings("rawtypes")
+    public void performNeighborNotificationOnTarget(IMixinWorldServer mixinWorld, BlockPos notifyPos, Block sourceBlock, BlockPos sourcePos,
+        IBlockState iblockstate) {
+        if (notifyPos.getX() == -1100 && notifyPos.getY() == 58 && notifyPos.getZ() == -496) {
```
Tsk.

Zidane commented on src/main/java/org/spongepowered/common/event/tracking/context/BlockTransaction.java:
```diff
@@ -0,0 +1,275 @@
+        return null;
+    }
+
+    static class TileEntityAdd extends BlockTransaction {
```
AddTileEntity

gabizou committed:
Lord of the Blocks: The Forbidden Tile

Fixed 0-tick pistons disappearing. There is still an issue somewhere down
the line of processing, whether it's with scheduled block events, or
neighbor notifications that ends up causing complex 0-tick pistons to get
the wrong block state. This is causing not so much of a duplication, but
irregular and non-deterministic block changes that otherwise would not
occur in vanilla. For single reactionary 0-tick piston contraptions that
do not passively trigger other 0-tick pistons, this branch is working. For
complex multi-reactionary 0-tick pistons that chain off one another, this
is still failing and potentially disappearing blocks or moving blocks when
they're not supposed to be moved.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Merge stable-7 changes into branch for development updates.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

liach commented:
finally!
